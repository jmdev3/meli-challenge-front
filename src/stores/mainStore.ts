import { createContext, useContext } from "react";
import { types, Instance, flow, getEnv, applySnapshot } from "mobx-state-tree";

const Item = types.model("Item", {
  id: "",
  title: "",
  price: types.model({
    currency: "",
    amount: types.number,
    decimals: types.number,
  }),
  picture: "",
  condition: "",
  free_shipping: types.boolean,
  sold_quantity: types.maybe(types.number),
  description: "",
});

export const MainStore = types
  .model("MainStore", {
    author: types.model({
      name: "",
      lastname: "",
    }),
    items: types.array(Item),
    categories: types.array(types.string),
  })
  .views((self) => {
    return {
      get selectedItem() {
        return self.items.length > 0 ? self.items[0] : null;
      },
    };
  })
  .actions((self) => {
    const searchItems = flow(function* (queryString: string) {
      const { api } = getEnv(self);

      try {
        self.items.clear();
        self.categories.clear();
        const response = yield api.searchItems({ search: queryString });
        applySnapshot(self.author, response.data.author);
        applySnapshot(self.categories, response.data.categories);
        applySnapshot(self.items, response.data.items);
      } catch (error) {
        alert("Hubo un error el buscar los productos!");
        console.log("> searchItems error: ", error);
      }
    });

    const getItemDetails = flow(function* (itemId: string) {
      const { api } = getEnv(self);

      try {
        self.items.clear();
        const response = yield api.getItem(itemId);
        applySnapshot(self.author, response.data.author);
        applySnapshot(self.categories, response.data.categories);
        applySnapshot(self.items, [response.data.item]);
      } catch (error) {
        alert("Hubo un error el buscar los detalles del producto!");
        console.log("> getItemDetails error: ", error);
      }
    });

    function clearStore() {
      self.items.clear();
      self.categories.clear();
    }

    return {
      searchItems,
      getItemDetails,
      clearStore,
    };
  });

export interface IItem extends Instance<typeof Item> {}
export interface IMainStore extends Instance<typeof MainStore> {}
const MainStoreContext = createContext<IMainStore>(null);
export const MainStoreProvider = MainStoreContext.Provider;
export const useMainStore = () => useContext<IMainStore>(MainStoreContext);
