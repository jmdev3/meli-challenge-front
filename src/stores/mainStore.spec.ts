import { MainStore } from "./mainStore";

describe("MainStore", () => {
  const mockedItem = {
    id: "item-id",
    title: "lenovo",
    price: {
      currency: "ARS",
      amount: 1000,
      decimals: 2,
    },
    picture: "http://http2.mlstatic.com/D_961352-MLA42764591253_072020-O.jpg",
    condition: "new",
    free_shipping: true,
    description: "item description",
    sold_quantity: 2,
  };

  const mockedCategories = ["ComputaciÃ³n", "Laptops y Accesorios"];

  const mockedAuthor = {
    name: "Juan Manuel",
    lastname: "Villarraza",
  };

  it("should init store as expected", () => {
    const store = MainStore.create({ author: { name: "", lastname: "" } });

    expect(store).toMatchSnapshot();
    expect(store.items).toHaveLength(0);
    expect(store.categories).toHaveLength(0);
    expect(store.author).toStrictEqual({
      name: "",
      lastname: "",
    });
  });

  describe("store actions", () => {
    it("should setItems as expected", () => {
      const store = MainStore.create({ author: { name: "", lastname: "" } });
      store.setItems([mockedItem]);
      const item = store.items[0];

      expect(item.id).toBe(mockedItem.id);
      expect(item.title).toBe(mockedItem.title);
      expect(item.price.currency).toBe(mockedItem.price.currency);
      expect(item.price.amount).toBe(mockedItem.price.amount);
      expect(item.price.decimals).toBe(mockedItem.price.decimals);
      expect(item.picture).toBe(mockedItem.picture);
      expect(item.condition).toBe(mockedItem.condition);
      expect(item.free_shipping).toBe(mockedItem.free_shipping);
      expect(item.sold_quantity).toBe(mockedItem.sold_quantity);
      expect(item.description).toBe(mockedItem.description);
    });

    it("should setCategories as expected", () => {
      const store = MainStore.create({ author: { name: "", lastname: "" } });
      store.setCategories(mockedCategories);

      expect(store.categories).toStrictEqual(mockedCategories);
    });

    it("should setAuthor as expected", () => {
      const store = MainStore.create({ author: { name: "", lastname: "" } });
      store.setAuthor(mockedAuthor);

      expect(store.author).toStrictEqual(mockedAuthor);
    });

    it("should clearStore as expected", () => {
      const store = MainStore.create({ author: { name: "", lastname: "" } });
      store.setItems([mockedItem]);
      store.setCategories(mockedCategories);
      store.setAuthor(mockedAuthor);
      store.clearStore();

      expect(store.items).toHaveLength(0);
      expect(store.categories).toHaveLength(0);
      expect(store.author).toStrictEqual({
        name: "Juan Manuel",
        lastname: "Villarraza",
      });
    });
  });
});
