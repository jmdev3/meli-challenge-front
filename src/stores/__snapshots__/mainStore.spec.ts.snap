// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainStore should init store as expected 1`] = `
Object {
  "author": Object {
    "lastname": "",
    "name": "",
  },
  "categories": Array [],
  "items": Array [],
}
`;

exports[`MainStore store actions should clearStore as expected 1`] = `
Object {
  "author": Object {
    "lastname": "Villarraza",
    "name": "Juan Manuel",
  },
  "categories": Array [],
  "items": Array [],
}
`;
